# Part 1

In [1]:
import numpy as np

In [2]:
class moonsOfJupiter():
    def __init__(self, input_line):
        self.position = self.parse_input(input_line)
        self.velocity = [0, 0, 0]
        
        self.dv = [0, 0, 0]
        
        self.x_period, self.y_period, self.z_period = 0, 0, 0
        self.vx_period, self.vy_period, self.vz_period = 0, 0, 0
        
        self.xhistory, self.yhistory, self.zhistory = set(), set(), set()
        self.vxhistory, self.vyhistory, self.vzhistory = set(), set(), set()

    def __repr__(self):
        return (self.position[0], self.position[1], self.position[2], 
                self.velocity[0], self.velocity[1], self.velocity[2])
    
    def parse_input(self, input_line):
        input_line = input_line[1:-2].split(', ')
        
        x = int(input_line[0][2:])
        y = int(input_line[1][2:])
        z = int(input_line[2][2:])
        
        return [x, y, z]
    
    def calculate_v_update(self, moon1, moon2, moon3):
        for i in range(3):
            self.dv[i] = self.calculate_increment(self.position[i], moon1.position[i]) + \
                         self.calculate_increment(self.position[i], moon2.position[i]) + \
                         self.calculate_increment(self.position[i], moon3.position[i])
    
    def calculate_increment(self, one, two):
        if (one<two):
            return 1
        elif(one>two):
            return -1
        else:
            return 0
        
    def update_pv(self):
        for i in range(3):
            self.velocity[i] += self.dv[i]
        for i in range(3):
            self.position[i] += self.velocity[i]
            
        self.dv = [0, 0, 0]
        
    def calculate_pe(self):
        pe = 0
        for i in range(3):
            pe += abs(self.position[i]) 
        return pe
    
    def calculate_ke(self):
        ke = 0
        for i in range(3):
            ke += abs(self.velocity[i]) 
        return ke
    
    def calculate_te(self):
        '''calculate total energy'''
        total_energy = self.calculate_pe() * self.calculate_ke()
        return total_energy
    
    def print_pv(self):
        print([self.position[0], self.position[1], self.position[2], self.velocity[0], self.velocity[1], self.velocity[2]])
        
    def check_histories(self, i):
        
        if not self.x_period:
            if self.position[0] not in self.xhistory:
                self.xhistory.add(self.position[0])
            else:
                self.x_period = i
                
        if not self.y_period:
            if self.position[1] not in self.yhistory:
                self.yhistory.add(self.position[1])
            else:
                self.y_period = i

        if not self.z_period:
            if self.position[2] not in self.zhistory:
                self.zhistory.add(self.position[2])
            else:
                self.z_period = i

        if not self.vx_period:
            if self.velocity[0] not in self.vxhistory:
                self.vxhistory.add(self.velocity[0])
            else:
                self.vx_period = i
                
        if not self.vy_period:
            if self.velocity[1] not in self.vyhistory:
                self.vyhistory.add(self.velocity[1])
            else:
                self.vy_period = i

        if not self.vz_period:
            if self.velocity[2] not in self.vzhistory:
                self.vzhistory.add(self.velocity[2])
            else:
                self.vz_period = i

In [3]:
input_file = 'day12_input.txt'
# input_file = 'day12_test_input.txt'
with open(input_file, 'r') as f:
    lines = f.readlines()
    
# Setting initial positions and velocities for the moons
io = moonsOfJupiter(lines[0])
europa = moonsOfJupiter(lines[1])
ganymede = moonsOfJupiter(lines[2])
callisto = moonsOfJupiter(lines[3])

In [4]:
# Let's simulate the trajectories!
for i in range(1000):
    io.calculate_v_update(europa, ganymede, callisto)
    europa.calculate_v_update(io, ganymede, callisto)
    ganymede.calculate_v_update(io, europa, callisto)
    callisto.calculate_v_update(io, europa, ganymede)
    
    io.update_pv()
    europa.update_pv()
    ganymede.update_pv()
    callisto.update_pv() 

In [5]:
print(io.calculate_te() + ganymede.calculate_te() + europa.calculate_te() + callisto.calculate_te())

183


# Part 2

In [250]:
input_file = 'day12_input.txt'
# input_file = 'day12_test_input.txt'
with open(input_file, 'r') as f:
    lines = f.readlines()
    
# Setting initial positions and velocities for the moons
io = moonsOfJupiter(lines[0])
europa = moonsOfJupiter(lines[1])
ganymede = moonsOfJupiter(lines[2])
callisto = moonsOfJupiter(lines[3])

io_initial = io.__repr__()
europa_initial = europa.__repr__()
ganymede_initial = ganymede.__repr__()
callisto_initial = callisto.__repr__()

In [251]:
io_period = [0]
europa_period = [0]
ganymede_period = [0]
callisto_period = [0]

In [252]:
io_periods = [[], [], [], [], [], []]
europa_periods = [[], [], [], [], [], []]
ganymede_periods = [[], [], [], [], [], []]
callisto_periods = [[], [], [], [], [], []]

In [253]:
def fillPeriods(periods, inputValue, refValue, i):
    
    for j, val in enumerate(periods):
        if((inputValue[j] == refValue[j])):
            periods[j].append(i)
    
    return periods

In [254]:
# Let's simulate the trajectories!
for i in range(1, 10000):
    io.calculate_v_update(europa, ganymede, callisto)
    europa.calculate_v_update(io, ganymede, callisto)
    ganymede.calculate_v_update(io, europa, callisto)
    callisto.calculate_v_update(io, europa, ganymede)
    
    io.update_pv()
    europa.update_pv()
    ganymede.update_pv()
    callisto.update_pv()
    
    if io.__repr__() == io_initial:
        io_period.append(i)
        
    if europa.__repr__() == europa_initial:
        europa_period.append(i)
        
    if ganymede.__repr__() == ganymede_initial:
        ganymede_period.append(i)
        
    if callisto.__repr__() == callisto_initial:
        callisto_period.append(i)

    io_periods = fillPeriods(io_periods, io.__repr__(), io_initial, i)
    europa_periods =fillPeriods(europa_periods, europa.__repr__(), europa_initial, i)
    ganymede_periods = fillPeriods(ganymede_periods, ganymede.__repr__(), ganymede_initial, i)
    callisto_periods = fillPeriods(callisto_periods, callisto.__repr__(), callisto_initial, i)

    

In [258]:
np.array(io_periods[0][1:]) - np.array(io_periods[0][0:-1])

array([   9,   56,   34,    1,   12,   66,   12,   28,    1,    3,   84,
          7,   11,    4,    1,    6,   86,  445,  135,  864,  337,   23,
        689,  324,  631,  100,   21,  131,   60,   77,   23,  354, 1344,
       2170,  465,  271])

In [238]:
# Io periods - 6, 28, 44, 3, 14, 44

In [ ]:
# Europa periods - 9, 28, 44, 9, 28, 44

In [248]:
europa_periods[4]

[5,
 14,
 23,
 28,
 33,
 42,
 51,
 56,
 61,
 70,
 79,
 84,
 89,
 98,
 107,
 112,
 117,
 126,
 135,
 140,
 145,
 154,
 163,
 168,
 173,
 182,
 191,
 196,
 201,
 210,
 219,
 224,
 229,
 238,
 247,
 252,
 257,
 266,
 275,
 280,
 285,
 294,
 303,
 308,
 313,
 322,
 331,
 336,
 341,
 350,
 359,
 364,
 369,
 378,
 387,
 392,
 397,
 406,
 415,
 420,
 425,
 434,
 443,
 448,
 453,
 462,
 471,
 476,
 481,
 490,
 499,
 504,
 509,
 518,
 527,
 532,
 537,
 546,
 555,
 560,
 565,
 574,
 583,
 588,
 593,
 602,
 611,
 616,
 621,
 630,
 639,
 644,
 649,
 658,
 667,
 672,
 677,
 686,
 695,
 700,
 705,
 714,
 723,
 728,
 733,
 742,
 751,
 756,
 761,
 770,
 779,
 784,
 789,
 798,
 807,
 812,
 817,
 826,
 835,
 840,
 845,
 854,
 863,
 868,
 873,
 882,
 891,
 896,
 901,
 910,
 919,
 924,
 929,
 938,
 947,
 952,
 957,
 966,
 975,
 980,
 985,
 994,
 1003,
 1008,
 1013,
 1022,
 1031,
 1036,
 1041,
 1050,
 1059,
 1064,
 1069,
 1078,
 1087,
 1092,
 1097,
 1106,
 1115,
 1120,
 1125,
 1134,
 1143,
 1148,
 1153,
 1

In [249]:
np.lcm.reduce([9, 28, 44, 9, 28, 44])

2772

In [196]:
(np.lcm.reduce(np.array(io_periods)), np.lcm.reduce(europa_periods), np.lcm.reduce(ganymede_periods), np.lcm.reduce(callisto_periods))

(1890, 441180, 1750014, 135)

### So the periods of the moons are 924, 2772, 2772, and 924 time steps respectively (accounting for zero-indexing). Let's then calculate the least common multiple of 924 and 2772 to figure out the period of the system -